File: PythonSandbox/src/leetcode/lc13_roman_to_integer.py
'''
13. Roman to Integer
https://leetcode.com/problems/roman-to-integer/

Symbol       Value
I             1
V             5
X             10
L             50
C             100
D             500
M             1000
'''

class Solution:
    '''
    Trick: go backwards instead of forwards.
    Let n = len(s)
    Time complexity: O(n), since n elements are iterated through.
    Space complexity: O(1), since tot stores the sum so far during the iteration of each char.
        (Assuming that rti map can be ignored, since that is necessary for this algorithm to work.)
    '''
    def romanToInt(self, s: str) -> int:
        rti = {
            'I' : 1,
            'V' : 5,
            'X' : 10,
            'L' : 50,
            'C' : 100,
            'D' : 500,
            'M' : 1000}
        
        tot = 0
        i = len(s)-1
        while i >= 0:
            #print("--- i={}, rnum:{}".format(i, s[i]))
            
            if i == 0:
                tot += rti[s[i]]
                break
            
            if rti[s[i]] > rti[s[i-1]]:
                # subtract
                diff = rti[s[i]] - rti[s[i-1]]
                tot += diff
                i = i-1
            else:
                # add
                tot += rti[s[i]]
            print("tot is: ", tot)
            i = i-1
                
        print("tot: ", tot)
        return tot
        
    def test1(self):
        #s = "MCMXCIV" # 1994
        #s = "XI"
        #s = "IX"
        #s = "IV"
        s = "LVIII" # 58
        tot = self.romanToInt(s)
        print("test1 tot: ", tot)
        
sol = Solution()
sol.test1()

                

File: JavaSandbox/src/leetcode/lc13_RomanToInteger.java
package leetcode;

import java.util.HashMap;

public class lc13_RomanToInteger {
    public HashMap<Character, Integer> rti;
    
    public lc13_RomanToInteger() {
        this.rti = new HashMap<>();
        rti.put('I', 1);
        rti.put('V', 5);
        rti.put('X', 10);
        rti.put('L', 50);
        rti.put('C', 100);
        rti.put('D', 500);
        rti.put('M', 1000);
    }
    
    public int romanToInt(String s) {
        int tot = 0;
        int i = s.length()-1;
        while(i >= 0) {
            if(i==0) {
                tot += rti.get(s.charAt(i));
                break;
            }
            if(rti.get(s.charAt(i)) > rti.get(s.charAt(i-1))) {
                int diff = rti.get(s.charAt(i))-rti.get(s.charAt(i-1));
                tot += diff;
                i -= 1;
            } else {
                tot += rti.get(s.charAt(i));
            }
            
            i -= 1;
        }
        
        return tot;
    }
}
